### ISSUE TITLE
thread 'main' panicked at 'assertion failed: self.active.edges.is_empty()

### ISSUE THREAD

**Issue by edap** [nical/lyon#599]
Hello, I am receiving this error.
thread 'main' panicked at 'assertion failed: self.active.edges.is_empty()', /home/my_user/.cargo/registry/src/github.com-1ecc6299db9ec823/lyon_tessellation-0.15.9/src/fill.rs:606:13

It happens randomly. I can't reproduce the bug on a small example, it happens randomly.

**Comment by nical** [comment]
Thanks for filing the issue. The tessellator is completely deterministic so If you manage to a set of inputs (path and options) that trigger the panic, I'll be able to make a reduced test case from it.

**Comment by edap** [comment]
Thanks for your reply Nicolas, I will try reproduce the polygon that crash locally.

**Comment by clarkmoody** [comment]
I've run into this issue. I'm not 100% sure, but the shape could be something like this:

```
A
|
|
B-------C
|       |
|       |
D-------E-------F

A.x = B.x = D.x
D.y = E.y = F.y
```

The instructions would be
```
moveTo(A)
lineTo(B)
lineTo(C)
lineTo(E)
lineTo(F)
lineTo(D)
fill()
```

My hypothesis is that the zero-width segments from `A-D` and `D-F` are triggering the error. I haven't been able to reliably reproduce.

**Comment by nical** [comment]
Thanks. I've tried with writing a test like this:

```
#[test]
fn issue_599() {
    let mut builder = Path::builder();

    builder.begin(point(1.0, 1.0));
    builder.line_to(point(1.0, 2.0));
    builder.line_to(point(2.0, 2.0));
    builder.line_to(point(2.0, 3.0));
    builder.line_to(point(3.0, 3.0));
    builder.line_to(point(1.0, 3.0));
    builder.close();

    let mut path = builder.build();

    test_path_with_rotations(path.clone(), 0.001, None);

    scale_path(&mut path, 100.0);

    test_path_with_rotations(path.clone(), 0.001, None);

    scale_path(&mut path, 100.0);

    test_path_with_rotations(path.clone(), 0.001, None);
}

```

Which tests thousands of rotations of an input paths which looks like the one you suggested, at different scales to try to flush out float precision issues.
I haven't managed to reproduce the bug yet but any hint will get us one step closer to finding a reproducing test case.

There is special code in the code to handle zero-width segments, but it's definitely a part that is difficult to get right because with finite precision, things are rarely never completely aligned/overlapping so it might be that combined with some other thing.

By the way, knowing the order of magnitude of the range of coordinates you are using can help.

**Comment by clarkmoody** [comment]
Coordinates were in the range of 0 to 500.

One more detail: this may have triggered on a case where the path was transformed with a 90-degree rotation and subsequent corrective scaling of one of the axes.

**Comment by OfficialURL** [comment]
I'm getting this same error after updating `lyon_path` to 0.17.3. The path that fails is the following:
```txt
" M -0.04457395 0.6926915 L -0.044577092 0.69268686 L 0.04457296 0.69263 L 0.044570256 0.69263405 L 0.04347263 0.676208 L 0.04339246 0.6761302 L -0.043470938 0.6761849 Z"
```

**Comment by nical** [comment]
Thanks a lot! I've investigated the test case and found the problem. It was an interesting one, requiring a few rather unusual things to line up. Fix coming soon.

**Comment by nical** [comment]
The fix is available in `lyon_tessellation 0.17.7`.

I'm going to be optimistic and close this, although there is no way to be 100% certain that it was the same root cause that caused all of you to hit this assertion. If you run into this assertion again with version 0.17.7 or higher, don't hesitate to comment here or open a new issue.
